File: src/commands/sync.ts
import { $ } from "bun";
import { findGwtRoot, checkGwtSetup, getWorktrees, debug } from "../core/repo";
import { resolveWorktree, selectWorktree } from "./cd";

interface SyncOptions {
  noFetch?: boolean;
}

export async function sync(name?: string, options: SyncOptions = {}): Promise<void> {
  const check = checkGwtSetup();
  if (!check.ok) {
    throw new Error(`Error: ${check.error}`);
  }

  const root = findGwtRoot()!;
  process.chdir(root);

  if (!options.noFetch) {
    console.log("Fetching remotes...");
    const fetch = await $`git fetch --all`.quiet().nothrow();
    if (fetch.exitCode !== 0) {
      console.error("Warning: Failed to fetch remotes");
    }
  }

  const worktrees = await getWorktrees();
  if (worktrees.length === 0) {
    throw new Error("No worktrees found");
  }

  const wt = name ? resolveWorktree(worktrees, name) : await selectWorktree(worktrees);

  debug("sync", { name: wt.name, path: wt.path, branch: wt.branch });
  console.log(`Syncing '${wt.name}'...`);
  const result = await $`git -C ${wt.path} pull --rebase`.quiet().nothrow();

  if (result.exitCode !== 0) {
    const stderr = result.stderr.toString().trim();
    throw new Error(`Error: Failed to sync\n${stderr}`);
  }

  const output = result.stdout.toString().trim();
  if (output) {
    console.log(output);
  }
  console.log(`Done! '${wt.name}' is up to date`);
}

File: src/sync.rs
use std::ffi::OsString;

use crate::AppResult;
use crate::arg_to_str;
use crate::cd::{resolve_worktree, select_worktree};
use crate::repo::{ensure_gwt_setup, get_worktrees, git, git_in_worktree};

pub fn run(args: &[OsString]) -> AppResult<()> {
    let (name, no_fetch) = parse_options(args)?;
    let root = ensure_gwt_setup()?;

    if !no_fetch {
        println!("Fetching remotes...");
        let fetch = git(["fetch", "--all"], Some(&root))?;
        if fetch.exit_code != 0 {
            eprintln!("Warning: Failed to fetch remotes");
        }
    }

    let worktrees = get_worktrees(&root)?;
    if worktrees.is_empty() {
        return Err("No worktrees found".to_string());
    }

    let worktree = if let Some(name) = name.as_deref() {
        resolve_worktree(&worktrees, name)?.clone()
    } else {
        select_worktree(&worktrees)?
    };

    println!("Syncing '{}'...", worktree.name);
    let result = git_in_worktree(&worktree.path, &["pull", "--rebase"])?;
    if result.exit_code != 0 {
        return Err(format!("Error: Failed to sync\n{}", result.stderr.trim()));
    }
    if !result.stdout.trim().is_empty() {
        println!("{}", result.stdout.trim());
    }
    println!("Done! '{}' is up to date", worktree.name);
    Ok(())
}

fn parse_options(args: &[OsString]) -> AppResult<(Option<String>, bool)> {
    let mut name = None;
    let mut no_fetch = false;
    for arg in args {
        match arg_to_str(arg)? {
            "-n" | "--no-fetch" => no_fetch = true,
            value if value.starts_with('-') => return Err(format!("Error: unknown sync option '{value}'")),
            value => {
                if name.is_some() {
                    return Err("Error: sync accepts at most one worktree name".to_string());
                }
                name = Some(value.to_string());
            }
        }
    }
    Ok((name, no_fetch))
}
